roachtest: cdc/tpcc-1000/rangefeed=true failed

SHA: https://github.com/cockroachdb/cockroach/commits/fc7e48295cd05f94fd2883498d96d91ad538e559

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1308263&tab=buildLog

```
The test failed on branch=master, cloud=gce:
	cluster.go:1516,cdc.go:747,cdc.go:135,cluster.go:1854,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1308263-cdc-tpcc-1000-rangefeed-true:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		l
		   1m32s       60           19.0            9.0  18253.6  34359.7  38654.7  38654.7 delivery
		   1m32s       60          159.0           76.6  27917.3  77309.4  77309.4  77309.4 newOrder
		   1m32s       60           14.0            9.4  19327.4  24696.1  34359.7  34359.7 orderStatus
		   1m32s       60          150.0           90.1  24696.1  45097.2  73014.4  77309.4 payment
		   1m32s       60            7.0            9.4  19327.4  33286.0  33286.0  33286.0 stockLevel
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		   1m33s       60            9.0            9.0  19327.4  34359.7  34359.7  34359.7 delivery
		   1m33s       60          172.9           77.6  24696.1  62277.0  77309.4  77309.4 newOrder
		   1m33s       60           25.0            9.6  20401.1  38654.7  40802.2  40802.2 orderStatus
		   1m33s       60          158.9           90.9  24696.1  49392.1  57982.1  64424.5 payment
		   1m33s       60           18.0            9.5  20401.1  31138.5  31138.5  31138.5 stockLevel
		: signal: killed
	cluster.go:1875,cdc.go:223,cdc.go:475,test.go:1251: unexpected status: failed

```

Hmm, not clear what happened here. TPCC got an error:

```
E190525 08:05:06.806037 1 workload/cli/run.go:428  error in newOrder: ERROR: communication error: rpc error: code = Canceled desc = context canceled (SQLSTATE 08006)
   1m28s       47            3.0            8.9  21474.8  42949.7  42949.7  42949.7 delivery
   1m28s       47           15.0           73.1  27917.3  55834.6  57982.1  57982.1 newOrder
   1m28s       47            0.0            9.4      0.0      0.0      0.0      0.0 orderStatus
   1m28s       47           42.0           88.4  28991.0  77309.4  85899.3  85899.3 payment
   1m28s       47            1.0            9.5  31138.5  31138.5  31138.5  31138.5 stockLevel
_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
   1m29s       47            9.0            8.9  23622.3  34359.7  34359.7  34359.7 delivery
   1m29s       47          164.8           74.1  30064.8  73014.4  73014.4  73014.4 newOrder
   1m29s       47            6.0            9.4  24696.1  36507.2  36507.2  36507.2 orderStatus
   1m29s       47          106.9           88.6  30064.8  68719.5  85899.3  85899.3 payment
   1m29s       47            7.0            9.4  30064.8  40802.2  40802.2  40802.2 stockLevel
E190525 08:05:08.904460 1 workload/cli/run.go:428  error in stockLevel: ERROR: rpc error: code = Unavailable desc = transport is closing (SQLSTATE XXUUU)
```

The workload runs with `--tolerate-errors`, though. I'm not sure why the job failed (which I think is what "unexpected status: failed" is telling me). It'd sure be nice if the actual job were logged in this case.

SHA: https://github.com/cockroachdb/cockroach/commits/056afc72397974ed9489bf11f16ac8e2fc281a67

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1341966&tab=buildLog

```
The test failed on branch=master, cloud=gce:
	cluster.go:1513,cdc.go:744,cdc.go:132,cluster.go:1851,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1341966-cdc-tpcc-1000-rangefeed-true:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		     9.6  19327.4  27917.3  30064.8  30064.8 delivery
		    1m7s     2491          346.0           65.0  26843.5  49392.1  49392.1  49392.1 newOrder
		    1m7s     2491           19.0            9.6  19327.4  23622.3  23622.3  23622.3 orderStatus
		    1m7s     2491          185.0           87.2  25769.8  64424.5  64424.5  64424.5 payment
		    1m7s     2491           17.0            9.9  19327.4  25769.8  38654.7  38654.7 stockLevel
		E190617 07:55:41.096484 1 workload/cli/run.go:425  error in payment: dial tcp 10.142.0.30:26257: connect: connection refused
		    1m8s     2738           13.0            9.7   5637.1  21474.8  24696.1  24696.1 delivery
		    1m8s     2738          221.3           67.3  26843.5  51539.6  51539.6  51539.6 newOrder
		    1m8s     2738           18.0            9.7   4160.7  10737.4  13958.6  13958.6 orderStatus
		    1m8s     2738          115.2           87.6  21474.8  45097.2  66572.0  66572.0 payment
		    1m8s     2738           17.0           10.0  10200.5  30064.8  36507.2  36507.2 stockLevel
		: signal: killed
	cluster.go:1872,cdc.go:220,cdc.go:472,test.go:1248: unexpected node event: 2: dead
	cluster.go:1035,context.go:127,cluster.go:1024,asm_amd64.s:522,panic.go:397,test.go:785,test.go:771,cluster.go:1872,cdc.go:220,cdc.go:472,test.go:1248: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1341966-cdc-tpcc-1000-rangefeed-true --oneshot --ignore-empty-nodes: exit status 1 4: skipped
		2: dead
		3: 4446
		1: 4261
		Error:  2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/9322e07476de447799c5d3011eb2874930ee2993

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1375546&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190706-1375546/cdc/tpcc-1000/rangefeed=true/run_1
	cluster.go:1724,cdc.go:748,cdc.go:136,cluster.go:2067,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1562393890-12-n4cpu16:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		   8.3  31138.5  38654.7  42949.7  42949.7 delivery
		    1m6s     2356          298.8           55.1  40802.2  49392.1  49392.1  49392.1 newOrder
		    1m6s     2356           16.0            9.6  24696.1  36507.2  40802.2  40802.2 orderStatus
		    1m6s     2356          171.9           79.5  34359.7  62277.0  64424.5  64424.5 payment
		    1m6s     2356           17.0            9.6  36507.2  42949.7  42949.7  42949.7 stockLevel
		E190706 10:19:57.969098 1 workload/cli/run.go:427  error in newOrder: dial tcp 10.142.0.126:26257: connect: connection refused
		    1m7s     2504           10.0            8.4  28991.0  34359.7  34359.7  34359.7 delivery
		    1m7s     2504          317.9           59.0  36507.2  49392.1  49392.1  49392.1 newOrder
		    1m7s     2504           20.0            9.7  18253.6  32212.3  32212.3  32212.3 orderStatus
		    1m7s     2504          188.9           81.1  32212.3  47244.6  64424.5  64424.5 payment
		    1m7s     2504           21.0            9.7  21474.8  38654.7  45097.2  45097.2 stockLevel
		: signal: killed
	cluster.go:2088,cdc.go:224,cdc.go:476,test_runner.go:681: unexpected node event: 2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/22d48caaa7d39efdcef7b3c87a99fc421e1473af

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1397412&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190720-1397412/cdc/tpcc-1000/rangefeed=true/run_1
	cluster.go:1726,cdc.go:748,cdc.go:136,cluster.go:2069,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1563603417-15-n4cpu16:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		.0          128.6  12884.9  18253.6  19327.4  19327.4 newOrder
		   1m36s     1442           16.0           16.4  13421.8  19327.4  19327.4  19327.4 orderStatus
		   1m36s     1442           62.0          162.5  14495.5  18253.6  19327.4  19327.4 payment
		   1m36s     1442            2.0           16.6  11811.2  12348.0  12348.0  12348.0 stockLevel
		E190720 09:14:34.829584 1 workload/cli/run.go:427  error in payment: dial tcp 10.128.0.87:26257: connect: connection refused
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		   1m37s     1611            0.0           16.4      0.0      0.0      0.0      0.0 delivery
		   1m37s     1611            7.0          127.4  15569.3  17179.9  17179.9  17179.9 newOrder
		   1m37s     1611            0.0           16.3      0.0      0.0      0.0      0.0 orderStatus
		   1m37s     1611            4.0          160.9  13958.6  18253.6  18253.6  18253.6 payment
		   1m37s     1611            2.0           16.5  15569.3  16643.0  16643.0  16643.0 stockLevel
		: signal: killed
	cluster.go:2090,cdc.go:224,cdc.go:476,test_runner.go:691: unexpected node event: 2: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/7dab0dcfd37c389af357c302c073b9611b5ada25

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1398203&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190721-1398203/cdc/tpcc-1000/rangefeed=true/run_1
	cluster.go:1726,cdc.go:748,cdc.go:136,cluster.go:2069,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1563689854-13-n4cpu16:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		    17.2  17179.9  17179.9  17179.9  17179.9 delivery
		   2m14s     1460           15.0          148.7  16106.1  19327.4  20401.1  20401.1 newOrder
		   2m14s     1460            5.0           18.3  16106.1  18253.6  18253.6  18253.6 orderStatus
		   2m14s     1460           24.0          170.8  17179.9  19327.4  20401.1  20401.1 payment
		   2m14s     1460           11.0           17.4  17179.9  19327.4  19327.4  19327.4 stockLevel
		E190721 09:27:55.183492 1 workload/cli/run.go:427  error in payment: dial tcp 10.128.0.99:26257: connect: connection refused
		   2m15s     1618            1.0           17.1  15569.3  15569.3  15569.3  15569.3 delivery
		   2m15s     1618           41.0          147.9  18253.6  20401.1  21474.8  21474.8 newOrder
		   2m15s     1618            4.0           18.2  16106.1  18253.6  18253.6  18253.6 orderStatus
		   2m15s     1618           52.0          169.9  18253.6  20401.1  20401.1  20401.1 payment
		   2m15s     1618            9.0           17.3  16106.1  17179.9  17179.9  17179.9 stockLevel
		: signal: killed
	cluster.go:2090,cdc.go:224,cdc.go:476,test_runner.go:691: unexpected node event: 3: dead

```

SHA: https://github.com/cockroachdb/cockroach/commits/86eab2ff0a1a4c2d9b5f7e7a45deda74c98c6c37

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1402541&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190724-1402541/cdc/tpcc-1000/rangefeed=true/run_1
	cluster.go:1726,cdc.go:748,cdc.go:136,cluster.go:2069,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1563949032-13-n4cpu16:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		   0.0 newOrder
		   1m10s     1097            0.0           12.2      0.0      0.0      0.0      0.0 orderStatus
		   1m10s     1097            0.0          121.2      0.0      0.0      0.0      0.0 payment
		   1m10s     1097            0.0           12.2      0.0      0.0      0.0      0.0 stockLevel
		E190724 09:32:48.415002 1 workload/cli/run.go:427  error in payment: dial tcp 10.128.0.124:26257: connect: connection refused
		   1m11s     1441            0.0           11.9      0.0      0.0      0.0      0.0 delivery
		   1m11s     1441            0.0           89.3      0.0      0.0      0.0      0.0 newOrder
		   1m11s     1441            0.0           12.0      0.0      0.0      0.0      0.0 orderStatus
		   1m11s     1441            0.0          119.5      0.0      0.0      0.0      0.0 payment
		   1m11s     1441            0.0           12.0      0.0      0.0      0.0      0.0 stockLevel
		E190724 09:32:49.478652 1 workload/cli/run.go:427  error in orderStatus: dial tcp 10.128.0.124:26257: connect: connection refused
		: signal: killed
	cluster.go:2090,cdc.go:224,cdc.go:476,test_runner.go:691: unexpected node event: 2: dead

```

--- a/pkg/storage/replica_application.go
+++ b/pkg/storage/replica_application.go
@@ -382,6 +382,19 @@ func (r *Replica) stageRaftCommand(
 			log.Fatal(ctx, err)
 		}
 	}
+
+	// Provide the command's corresponding logical operations to the Replica's
+	// rangefeed. Only do so if the WriteBatch is non-nil, in which case the
+	// rangefeed requires there to be a corresponding logical operation log or
+	// it will shut down with an error. If the WriteBatch is nil then we expect
+	// the logical operation log to also be nil. We don't want to trigger a
+	// shutdown of the rangefeed in that situation, so we don't pass anything to
+	// the rangefed. If no rangefeed is running at all, this call will be a noop.
+	if cmd.raftCmd.WriteBatch != nil {
+		r.handleLogicalOpLogRaftMuLocked(ctx, cmd.raftCmd.LogicalOpLog, batch)
+	} else if cmd.raftCmd.LogicalOpLog != nil {
+		log.Fatalf(ctx, "non-nil logical op log with nil write batch: %v", cmd.raftCmd)
+	}
 }
 
 func checkForcedErr(

--- a/pkg/storage/replica_application.go
+++ b/pkg/storage/replica_application.go
@@ -154,23 +154,63 @@ func (r *Replica) retrieveLocalProposals(ctx context.Context, b *cmdAppBatch) {
 	// Copy stats as it gets updated in-place in applyRaftCommandToBatch.
 	b.replicaState.Stats = &b.stats
 	*b.replicaState.Stats = *r.mu.state.Stats
+	// Assign all the local proposals first then delete all of them from the map
+	// in a second pass. This ensures that we retrieve all proposals correctly
+	// even if the batch has multiple entries for the same proposal, in which
+	// case the proposal was reproposed (either under its original or a new
+	// MaxLeaseIndex) which we handle in a second pass below.
+	var anyLocal bool
 	var it cmdAppCtxBufIterator
-	haveProposalQuota := r.mu.proposalQuota != nil
 	for ok := it.init(&b.cmdBuf); ok; ok = it.next() {
 		cmd := it.cur()
 		cmd.proposal = r.mu.proposals[cmd.idKey]
 		if cmd.proposedLocally() {
 			// We initiated this command, so use the caller-supplied context.
 			cmd.ctx = cmd.proposal.ctx
-			delete(r.mu.proposals, cmd.idKey)
-			// At this point we're not guaranteed to have proposalQuota initialized,
-			// the same is true for quotaReleaseQueues. Only queue the proposal's
-			// quota for release if the proposalQuota is initialized.
-			if haveProposalQuota {
-				r.mu.quotaReleaseQueue = append(r.mu.quotaReleaseQueue, cmd.proposal.quotaSize)
-			}
+			anyLocal = true
 		} else {
 			cmd.ctx = ctx
+		}
+	}
+	if !anyLocal {
+		// Fast-path.
+		return
+	}
+	for ok := it.init(&b.cmdBuf); ok; ok = it.next() {
+		cmd := it.cur()
+		if !cmd.proposedLocally() {
+			continue
+		}
+		if cmd.raftCmd.MaxLeaseIndex != cmd.proposal.command.MaxLeaseIndex {
+			// If this entry does not have the most up-to-date view of the
+			// corresponding proposal's maximum lease index then the proposal
+			// must have been reproposed with a higher lease index. (see
+			// tryReproposeWithNewLeaseIndex). In that case, there's a newer
+			// version of the proposal in the pipeline, so don't remove the
+			// proposal from the map. We expect this entry to be rejected by
+			// checkForcedErr.
+			continue
+		}
+		// Delete the proposal from the proposals map. There may be reproposals
+		// of the proposal in the pipeline, but those will all have the same max
+		// lease index, meaning that they will all be rejected after this entry
+		// applies (successfully or otherwise). If tryReproposeWithNewLeaseIndex
+		// picks up the proposal on failure, it will re-add the proposal to the
+		// proposal map, but this won't affect anything in this cmdAppBatch.
+		//
+		// While here, add the proposal's quota size to the quota release queue.
+		// We check the proposal map again first to avoid double free-ing quota
+		// when reproposals from the same proposal end up in the same entry
+		// application batch.
+		if _, ok := r.mu.proposals[cmd.idKey]; !ok {
+			continue
+		}
+		delete(r.mu.proposals, cmd.idKey)
+		// At this point we're not guaranteed to have proposalQuota initialized,
+		// the same is true for quotaReleaseQueues. Only queue the proposal's
+		// quota for release if the proposalQuota is initialized.
+		if r.mu.proposalQuota != nil {
+			r.mu.quotaReleaseQueue = append(r.mu.quotaReleaseQueue, cmd.proposal.quotaSize)
 		}
 	}
 }
@@ -856,7 +896,7 @@ func (r *Replica) applyCmdAppBatch(
 		}
 		cmd.replicatedResult().SuggestedCompactions = nil
 		isNonTrivial := batchIsNonTrivial && it.isLast()
-		if errExpl, err = r.handleRaftCommandResult(ctx, cmd, isNonTrivial,
+		if errExpl, err = r.handleRaftCommandResult(cmd.ctx, cmd, isNonTrivial,
 			b.replicaState.UsingAppliedStateKey); err != nil {
 			return errExpl, err
 		}

--- a/pkg/storage/replica_application_result.go
+++ b/pkg/storage/replica_application_result.go
@@ -160,7 +160,11 @@ func (r *Replica) handleRaftCommandResult(
 ) (errExpl string, err error) {
 	// Set up the local result prior to handling the ReplicatedEvalResult to
 	// give testing knobs an opportunity to inspect it.
-	r.prepareLocalResult(cmd.ctx, cmd)
+	r.prepareLocalResult(ctx, cmd)
+	if log.ExpensiveLogEnabled(ctx, 2) {
+		log.VEvent(ctx, 2, cmd.localResult.String())
+	}
+
 	// Handle the ReplicatedEvalResult, executing any side effects of the last
 	// state machine transition.
 	//
@@ -168,7 +172,7 @@ func (r *Replica) handleRaftCommandResult(
 	// before notifying a potentially waiting client.
 	clearTrivialReplicatedEvalResultFields(cmd.replicatedResult(), usingAppliedStateKey)
 	if isNonTrivial {
-		r.handleComplexReplicatedEvalResult(cmd.ctx, *cmd.replicatedResult())
+		r.handleComplexReplicatedEvalResult(ctx, *cmd.replicatedResult())
 	} else if !cmd.replicatedResult().Equal(storagepb.ReplicatedEvalResult{}) {
 		log.Fatalf(ctx, "failed to handle all side-effects of ReplicatedEvalResult: %v",
 			cmd.replicatedResult())
@@ -186,13 +190,13 @@ func (r *Replica) handleRaftCommandResult(
 	}
 
 	if cmd.localResult != nil {
-		r.handleLocalEvalResult(cmd.ctx, *cmd.localResult)
+		r.handleLocalEvalResult(ctx, *cmd.localResult)
 	}
-	r.finishRaftCommand(cmd.ctx, cmd)
+	r.finishRaftCommand(ctx, cmd)
 	switch cmd.e.Type {
 	case raftpb.EntryNormal:
 		if cmd.replicatedResult().ChangeReplicas != nil {
-			log.Fatalf(cmd.ctx, "unexpected replication change from command %s", &cmd.raftCmd)
+			log.Fatalf(ctx, "unexpected replication change from command %s", &cmd.raftCmd)
 		}
 	case raftpb.EntryConfChange:
 		if cmd.replicatedResult().ChangeReplicas == nil {
@@ -307,6 +311,10 @@ func (r *Replica) handleComplexReplicatedEvalResult(
 // result will zero-out the struct to ensure that is has properly performed all
 // of the implied side-effects.
 func (r *Replica) prepareLocalResult(ctx context.Context, cmd *cmdAppCtx) {
+	if !cmd.proposedLocally() {
+		return
+	}
+
 	var pErr *roachpb.Error
 	if filter := r.store.cfg.TestingKnobs.TestingPostApplyFilter; filter != nil {
 		var newPropRetry int
@@ -328,36 +336,90 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *cmdAppCtx) {
 		pErr = cmd.forcedErr
 	}
 
-	if cmd.proposedLocally() {
-		if cmd.proposalRetry != proposalNoReevaluation && pErr == nil {
-			log.Fatalf(ctx, "proposal with nontrivial retry behavior, but no error: %+v", cmd.proposal)
-		}
-		if pErr != nil {
-			// A forced error was set (i.e. we did not apply the proposal,
-			// for instance due to its log position) or the Replica is now
-			// corrupted.
-			// If proposalRetry is set, we don't also return an error, as per the
-			// proposalResult contract.
-			if cmd.proposalRetry == proposalNoReevaluation {
-				cmd.response.Err = pErr
-			}
-		} else if cmd.proposal.Local.Reply != nil {
-			cmd.response.Reply = cmd.proposal.Local.Reply
-		} else {
-			log.Fatalf(ctx, "proposal must return either a reply or an error: %+v", cmd.proposal)
-		}
-		cmd.response.Intents = cmd.proposal.Local.DetachIntents()
-		cmd.response.EndTxns = cmd.proposal.Local.DetachEndTxns(pErr != nil)
-		if pErr == nil {
-			cmd.localResult = cmd.proposal.Local
-		}
+	if cmd.proposalRetry != proposalNoReevaluation && pErr == nil {
+		log.Fatalf(ctx, "proposal with nontrivial retry behavior, but no error: %+v", cmd.proposal)
 	}
-	if pErr != nil && cmd.localResult != nil {
+	if pErr != nil {
+		// A forced error was set (i.e. we did not apply the proposal,
+		// for instance due to its log position) or the Replica is now
+		// corrupted.
+		switch cmd.proposalRetry {
+		case proposalNoReevaluation:
+			cmd.response.Err = pErr
+		case proposalIllegalLeaseIndex:
+			// If we failed to apply at the right lease index, try again with a
+			// new one. This is important for pipelined writes, since they don't
+			// have a client watching to retry, so a failure to eventually apply
+			// the proposal would be a user-visible error.
+			pErr = r.tryReproposeWithNewLeaseIndex(ctx, cmd)
+			if pErr != nil {
+				cmd.response.Err = pErr
+			} else {
+				// Unbind the entry's local proposal because we just succeeded
+				// in reproposing it and we don't want to acknowledge the client
+				// yet.
+				cmd.proposal = nil
+				return
+			}
+		default:
+			panic("unexpected")
+		}
+	} else if cmd.proposal.Local.Reply != nil {
+		cmd.response.Reply = cmd.proposal.Local.Reply
+	} else {
+		log.Fatalf(ctx, "proposal must return either a reply or an error: %+v", cmd.proposal)
+	}
+	cmd.response.Intents = cmd.proposal.Local.DetachIntents()
+	cmd.response.EndTxns = cmd.proposal.Local.DetachEndTxns(pErr != nil)
+	if pErr == nil {
+		cmd.localResult = cmd.proposal.Local
+	} else if cmd.localResult != nil {
 		log.Fatalf(ctx, "shouldn't have a local result if command processing failed. pErr: %s", pErr)
 	}
-	if log.ExpensiveLogEnabled(ctx, 2) {
-		log.VEvent(ctx, 2, cmd.localResult.String())
+}
+
+// tryReproposeWithNewLeaseIndex is used by prepareLocalResult to repropose
+// commands that have gotten an illegal lease index error, and that we know
+// could not have applied while their lease index was valid (that is, we
+// observed all applied entries between proposal and the lease index becoming
+// invalid, as opposed to skipping some of them by applying a snapshot).
+//
+// It is not intended for use elsewhere and is only a top-level function so that
+// it can avoid the below_raft_protos check. Returns a nil error if the command
+// has already been successfully applied or has been reproposed here or by a
+// different entry for the same proposal that hit an illegal lease index error.
+func (r *Replica) tryReproposeWithNewLeaseIndex(
+	ctx context.Context, cmd *cmdAppCtx,
+) *roachpb.Error {
+	// Note that we don't need to validate anything about the proposal's
+	// lease here - if we got this far, we know that everything but the
+	// index is valid at this point in the log.
+	p := cmd.proposal
+	if p.applied || cmd.raftCmd.MaxLeaseIndex != p.command.MaxLeaseIndex {
+		// If the command associated with this rejected raft entry already
+		// applied then we don't want to repropose it. Doing so could lead
+		// to duplicate application of the same proposal.
+		//
+		// Similarly, if the command associated with this rejected raft
+		// entry has a different (larger) MaxLeaseIndex than the one we
+		// decoded from the entry itself, the command must have already
+		// been reproposed (this can happen if there are multiple copies
+		// of the command in the logs; see TestReplicaRefreshMultiple).
+		// We must not create multiple copies with multiple lease indexes,
+		// so don't repropose it again. This ensures that at any time,
+		// there is only up to a single lease index that has a chance of
+		// succeeding in the Raft log for a given command.
+		return nil
 	}
+	// Some tests check for this log message in the trace.
+	log.VEventf(ctx, 2, "retry: proposalIllegalLeaseIndex")
+	maxLeaseIndex, pErr := r.propose(ctx, p)
+	if pErr != nil {
+		log.Warningf(ctx, "failed to repropose with new lease index: %s", pErr)
+		return pErr
+	}
+	log.VEventf(ctx, 2, "reproposed command %x at maxLeaseIndex=%d", cmd.idKey, maxLeaseIndex)
+	return nil
 }
 
 // finishRaftCommand is called after a command's side effects have been applied
@@ -395,17 +457,6 @@ func (r *Replica) finishRaftCommand(ctx context.Context, cmd *cmdAppCtx) {
 	}
 
 	if cmd.proposedLocally() {
-		// If we failed to apply at the right lease index, try again with
-		// a new one. This is important for pipelined writes, since they
-		// don't have a client watching to retry, so a failure to
-		// eventually apply the proposal would be a uservisible error.
-		// TODO(nvanbenschoten): This reproposal is not tracked by the
-		// quota pool. We should fix that.
-		if cmd.proposalRetry == proposalIllegalLeaseIndex &&
-			r.tryReproposeWithNewLeaseIndex(cmd.proposal) {
-			return
-		}
-		// Otherwise, signal the command's status to the client.
 		cmd.proposal.finishApplication(cmd.response)
 	} else if cmd.response.Err != nil {
 		log.VEventf(ctx, 1, "applying raft command resulted in error: %s", cmd.response.Err)

--- a/pkg/storage/replica_application_result.go
+++ b/pkg/storage/replica_application_result.go
@@ -363,18 +363,6 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *cmdAppCtx) {
 // finishRaftCommand is called after a command's side effects have been applied
 // in order to acknowledge clients and release latches.
 func (r *Replica) finishRaftCommand(ctx context.Context, cmd *cmdAppCtx) {
-
-	// Provide the command's corresponding logical operations to the
-	// Replica's rangefeed. Only do so if the WriteBatch is nonnil,
-	// otherwise it's valid for the logical op log to be nil, which
-	// would shut down all rangefeeds. If no rangefeed is running,
-	// this call will be a noop.
-	if cmd.raftCmd.WriteBatch != nil {
-		r.handleLogicalOpLogRaftMuLocked(ctx, cmd.raftCmd.LogicalOpLog)
-	} else if cmd.raftCmd.LogicalOpLog != nil {
-		log.Fatalf(ctx, "nonnil logical op log with nil write batch: %v", cmd.raftCmd)
-	}
-
 	// When set to true, recomputes the stats for the LHS and RHS of splits and
 	// makes sure that they agree with the state's range stats.
 	const expensiveSplitAssertion = false

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -87,6 +87,11 @@ type ProposalData struct {
 	// cache and release latches.
 	ec endCmds
 
+	// applied is set when the a command finishes application. It is used to
+	// avoid reproposing a failed proposal if an earlier version of the same
+	// proposal succeeded in applying.
+	applied bool
+
 	// doneCh is used to signal the waiting RPC handler (the contents of
 	// proposalResult come from LocalEvalResult).
 	//
@@ -122,11 +127,25 @@ type ProposalData struct {
 // is canceled, it won't be listening to this done channel, and so it can't be
 // counted on to invoke endCmds itself.)
 func (proposal *ProposalData) finishApplication(pr proposalResult) {
+	if proposal.applied {
+		// If the command already applied then we shouldn't be "finishing" its
+		// application again because it should only be able to apply successfully
+		// once. We expect that when any reproposal for the same command attempts
+		// to apply it will be rejected by the below raft lease sequence or lease
+		// index check in checkForcedErr.
+		if pr.Err != nil {
+			return
+		}
+		log.Fatalf(proposal.ctx,
+			"command already applied: %+v; unexpected successful result: %+v", proposal, pr)
+	}
+	proposal.applied = true
 	proposal.ec.done(proposal.Request, pr.Reply, pr.Err)
+	proposal.signalProposalResult(pr)
 	if proposal.sp != nil {
 		tracing.FinishSpan(proposal.sp)
+		proposal.sp = nil
 	}
-	proposal.signalProposalResult(pr)
 }
 
 // returnProposalResult signals proposal.doneCh with the proposal result if it

--- a/pkg/storage/replica_raft.go
+++ b/pkg/storage/replica_raft.go
@@ -1404,56 +1404,6 @@ func (m lastUpdateTimesMap) isFollowerActive(
 	return now.Sub(lastUpdateTime) <= MaxQuotaReplicaLivenessDuration
 }
 
-// tryReproposeWithNewLeaseIndex is used by processRaftCommand to
-// repropose commands that have gotten an illegal lease index error,
-// and that we know could not have applied while their lease index was
-// valid (that is, we observed all applied entries between proposal
-// and the lease index becoming invalid, as opposed to skipping some
-// of them by applying a snapshot).
-//
-// It is not intended for use elsewhere and is only a top-level
-// function so that it can avoid the below_raft_protos check. Returns
-// true if the command has been successfully reproposed (not
-// necessarily by this method! But if this method returns true, the
-// command will be in the local proposals map).
-func (r *Replica) tryReproposeWithNewLeaseIndex(proposal *ProposalData) bool {
-	// Note that we don't need to validate anything about the proposal's
-	// lease here - if we got this far, we know that everything but the
-	// index is valid at this point in the log.
-	r.mu.Lock()
-	if proposal.command.MaxLeaseIndex > r.mu.state.LeaseAppliedIndex {
-		// If the command's MaxLeaseIndex is greater than the
-		// LeaseAppliedIndex, it must have already been reproposed (this
-		// can happen if there are multiple copies of the command in the
-		// logs; see TestReplicaRefreshMultiple). We must not create
-		// multiple copies with multiple lease indexes, so don't repropose
-		// it again. This ensures that at any time, there is only up to a
-		// single lease index that has a chance of succeeding in the Raft
-		// log for a given command.
-		//
-		// Note that the caller has already removed the current version of
-		// the proposal from the pending proposals map. We must re-add it
-		// since it's still pending.
-		log.VEventf(proposal.ctx, 2, "skipping reproposal, already reproposed at index %d",
-			proposal.command.MaxLeaseIndex)
-		r.mu.proposals[proposal.idKey] = proposal
-		r.mu.Unlock()
-		return true
-	}
-	r.mu.Unlock()
-
-	// Some tests check for this log message in the trace.
-	log.VEventf(proposal.ctx, 2, "retry: proposalIllegalLeaseIndex")
-	if _, pErr := r.propose(proposal.ctx, proposal); pErr != nil {
-		// TODO(nvanbenschoten): Returning false here isn't ok. It will result
-		// in a proposal returning without a response or an error, which
-		// triggers a panic higher up in the stack. We need to fix this.
-		log.Warningf(proposal.ctx, "failed to repropose with new lease index: %s", pErr)
-		return false
-	}
-	return true
-}
-
 // maybeAcquireSnapshotMergeLock checks whether the incoming snapshot subsumes
 // any replicas and, if so, locks them for subsumption. See acquireMergeLock
 // for details about the lock itself.

--- a/pkg/storage/replica_rangefeed.go
+++ b/pkg/storage/replica_rangefeed.go
@@ -405,9 +405,13 @@ func (r *Replica) numRangefeedRegistrations() int {
 }
 
 // handleLogicalOpLogRaftMuLocked passes the logical op log to the active
-// rangefeed, if one is running. No-op if a rangefeed is not active. Requires
+// rangefeed, if one is running. The method accepts a reader, which is used to
+// look up the values associated with key-value writes in the log before handing
+// them to the rangefeed processor. No-op if a rangefeed is not active. Requires
 // raftMu to be locked.
-func (r *Replica) handleLogicalOpLogRaftMuLocked(ctx context.Context, ops *storagepb.LogicalOpLog) {
+func (r *Replica) handleLogicalOpLogRaftMuLocked(
+	ctx context.Context, ops *storagepb.LogicalOpLog, reader engine.Reader,
+) {
 	p := r.getRangefeedProcessor()
 	if p == nil {
 		return
@@ -429,8 +433,7 @@ func (r *Replica) handleLogicalOpLogRaftMuLocked(ctx context.Context, ops *stora
 	}
 
 	// When reading straight from the Raft log, some logical ops will not be
-	// fully populated. Read from the engine (under raftMu) to populate all
-	// fields.
+	// fully populated. Read from the Reader to populate all fields.
 	for _, op := range ops.Ops {
 		var key []byte
 		var ts hlc.Timestamp
@@ -450,12 +453,12 @@ func (r *Replica) handleLogicalOpLogRaftMuLocked(ctx context.Context, ops *stora
 			panic(fmt.Sprintf("unknown logical op %T", t))
 		}
 
-		// Read the value directly from the Engine. This is performed in the
+		// Read the value directly from the Reader. This is performed in the
 		// same raftMu critical section that the logical op's corresponding
 		// WriteBatch is applied, so the value should exist.
-		val, _, err := engine.MVCCGet(ctx, r.Engine(), key, ts, engine.MVCCGetOptions{Tombstones: true})
+		val, _, err := engine.MVCCGet(ctx, reader, key, ts, engine.MVCCGetOptions{Tombstones: true})
 		if val == nil && err == nil {
-			err = errors.New("value missing in engine")
+			err = errors.New("value missing in reader")
 		}
 		if err != nil {
 			r.disconnectRangefeedWithErr(p, roachpb.NewErrorf(

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -7849,6 +7849,88 @@ func TestReplicaRefreshMultiple(t *testing.T) {
 	}
 }
 
+// TestReplicaReproposalWithNewLeaseIndexError tests an interaction where a
+// proposal is rejected beneath raft due an illegal lease index error and then
+// hits an error when being reproposed. The expectation is that this error
+// manages to make its way back to the client.
+func TestReplicaReproposalWithNewLeaseIndexError(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	var tc testContext
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+	tc.Start(t, stopper)
+
+	type magicKey struct{}
+	magicCtx := context.WithValue(ctx, magicKey{}, "foo")
+
+	var c int32 // updated atomically
+	tc.repl.mu.Lock()
+	tc.repl.mu.proposalBuf.testing.leaseIndexFilter = func(p *ProposalData) (indexOverride uint64, _ error) {
+		if v := p.ctx.Value(magicKey{}); v != nil {
+			curAttempt := atomic.AddInt32(&c, 1)
+			switch curAttempt {
+			case 1:
+				// This is the first time the command is being given a max lease
+				// applied index. Set the index to that of the recently applied
+				// write. Two requests can't have the same lease applied index,
+				// so this will cause it to be rejected beneath raft with an
+				// illegal lease index error.
+				wrongLeaseIndex := uint64(1)
+				return wrongLeaseIndex, nil
+			case 2:
+				// This is the second time the command is being given a max
+				// lease applied index, which should be after the command was
+				// rejected beneath raft. Return an error. We expect this error
+				// to propagate up through tryReproposeWithNewLeaseIndex and
+				// make it back to the client.
+				return 0, errors.New("boom")
+			default:
+				// Unexpected. Asserted against below.
+				return 0, nil
+			}
+		}
+		return 0, nil
+	}
+	tc.repl.mu.Unlock()
+
+	// Perform a few writes to advance the lease applied index.
+	const initCount = 3
+	key := roachpb.Key("a")
+	for i := 0; i < initCount; i++ {
+		iArg := incrementArgs(key, 1)
+		if _, pErr := tc.SendWrapped(&iArg); pErr != nil {
+			t.Fatal(pErr)
+		}
+	}
+
+	// Perform a write that will first hit an illegal lease index error and
+	// will then hit the injected error when we attempt to repropose it.
+	var ba roachpb.BatchRequest
+	iArg := incrementArgs(key, 10)
+	ba.Add(&iArg)
+	if _, pErr := tc.Sender().Send(magicCtx, ba); pErr == nil {
+		t.Fatal("expected a non-nil error")
+	} else if !testutils.IsPError(pErr, "boom") {
+		t.Fatalf("unexpected error: %v", pErr)
+	}
+	// The command should have picked a new max lease index exactly twice.
+	if exp, act := int32(2), atomic.LoadInt32(&c); exp != act {
+		t.Fatalf("expected %d proposals, got %d", exp, act)
+	}
+
+	// The command should not have applied.
+	gArgs := getArgs(key)
+	if reply, pErr := tc.SendWrapped(&gArgs); pErr != nil {
+		t.Fatal(pErr)
+	} else if v, err := reply.(*roachpb.GetResponse).Value.GetInt(); err != nil {
+		t.Fatal(err)
+	} else if v != initCount {
+		t.Fatalf("expected value of %d, found %d", initCount, v)
+	}
+}
+
 // TestGCWithoutThreshold validates that GCRequest only declares the threshold
 // key if it is subject to change, and that it does not access this key if it
 // does not declare them.
@@ -7958,10 +8040,10 @@ func TestFailureToProcessCommandClearsLocalResult(t *testing.T) {
 	if err := testutils.MatchInOrder(formatted,
 		// The first proposal is rejected.
 		"retry proposal.*applied at lease index.*but required",
-		// The LocalResult is nil. This is the important part for this test.
-		"LocalResult: nil",
 		// The request will be re-evaluated.
 		"retry: proposalIllegalLeaseIndex",
+		// The LocalResult is nil. This is the important part for this test.
+		"LocalResult: nil",
 		// Re-evaluation succeeds and one txn is to be updated.
 		"LocalResult \\(reply.*#updated txns: 1",
 	); err != nil {

--- a/pkg/storage/testing_knobs.go
+++ b/pkg/storage/testing_knobs.go
@@ -56,6 +56,7 @@ type StoreTestingKnobs struct {
 
 	// TestingPostApplyFilter is called after a command is applied to
 	// rocksdb but before in-memory side effects have been processed.
+	// It is only called on the replica the proposed the command.
 	TestingPostApplyFilter storagebase.ReplicaApplyFilter
 
 	// TestingResponseFilter is called after the replica processes a

